overlord/configstate/hooks.go

```diff
@@ -133,3 +108,97 @@ func (h *configureHandler) Done() error {
 func (h *configureHandler) Error(err error) (bool, error) {
 	return false, nil
 }
+
+// defaultConfigureHandler is the handler for the default-configure hook.
+type defaultConfigureHandler struct {
+	context *hookstate.Context
+}
+
+func newDefaultConfigureHandler(context *hookstate.Context) hookstate.Handler {
+	return &defaultConfigureHandler{context: context}
+}
+
+// Before is called by the HookManager before the default-configure hook is run.
+func (h *defaultConfigureHandler) Before() error {
+	h.context.Lock()
+	defer h.context.Unlock()
+
+	tr := ContextTransaction(h.context)
+
+	instanceName := h.context.InstanceName()
+	st := h.context.State()
+	info, err := snapstate.CurrentInfo(st, instanceName)
+	if err != nil {
+		return err
+	}
+	hasDefaultConfigureHook := info.Hooks["default-configure"] != nil
+	hasConfigureHook := info.Hooks["configure"] != nil
+
+	// default-configure hook cannot be used without configure hook, because it is only intended
+	// as an extension of the configure hook that provides additional configuration support
+	if hasDefaultConfigureHook && !hasConfigureHook {
+		// this scenario should be prevented by the snap checker
+		return fmt.Errorf("cannot use default-configure hook for snap %q, no configure hook", instanceName)
```

Comment:
@pedronis, I though to keep to enforce coherent behaviour also without snap check e.g. for testing with mocking etc.
Would like to know your preference?